File: src/main/java/fr/lewon/bot/runner/util/BotTaskScheduler.java
package fr.lewon.bot.runner.util;

import fr.lewon.bot.runner.bot.task.BotTask;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.TaskScheduler;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ScheduledFuture;

@Component
public class BotTaskScheduler {

    @Autowired
    private TaskScheduler scheduler;

    // A map for keeping scheduled tasks
    private Map<BotTask, ScheduledFuture<?>> jobsMap = new HashMap<>();

    public void cancelTaskAutoExecution(BotTask task) {
        this.jobsMap.remove(task).cancel(true);
    }

    public void cancelTaskAutoExecution(List<BotTask> tasks) {
        tasks.forEach(this::cancelTaskAutoExecution);
    }

    public void startTaskAutoExecution(BotTask task) {
        this.jobsMap.put(task, this.scheduler.schedule(task, task));
    }

    public void startTaskAutoExecution(List<BotTask> tasks) {
        tasks.forEach(this::startTaskAutoExecution);
    }

    public List<BotTask> getTasks() {
        return new ArrayList<>(this.jobsMap.keySet());
    }
}

File: src/main/java/fr/lewon/bot/runner/util/BotTaskScheduler.kt
package fr.lewon.bot.runner.util

import fr.lewon.bot.runner.bot.task.BotTask
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.scheduling.TaskScheduler
import org.springframework.stereotype.Component
import java.util.concurrent.ScheduledFuture

@Component
class BotTaskScheduler {

    @Autowired
    private lateinit var scheduler: TaskScheduler

    private val jobsMap: MutableMap<BotTask, ScheduledFuture<*>> = HashMap()

    val tasks: List<BotTask>
        get() = ArrayList(this.jobsMap.keys)

    fun cancelTaskAutoExecution(task: BotTask) {
        this.jobsMap.remove(task)?.cancel(true)
    }

    fun cancelTaskAutoExecution(tasks: List<BotTask>) {
        tasks.forEach { t -> this.cancelTaskAutoExecution(t) }
    }

    fun startTaskAutoExecution(task: BotTask) {
        this.scheduler.schedule(task, task)?.let { this.jobsMap[task] = it }
    }

    fun startTaskAutoExecution(tasks: List<BotTask>) {
        tasks.forEach { t -> this.startTaskAutoExecution(t) }
    }
}
